docs/reference/builder.md

```diff
@@ -954,6 +954,125 @@ For example:
+elsewhere.  For example, consider this Dockerfile:
+
+```
+1 FROM busybox
+2 USER ${user:-some_user}
+3 ARG user
+4 USER $user
+...
+```
+A user builds this file by calling:
+
+```
+$ docker build --build-arg user=what_user Dockerfile
+```
+
+The `USER` at line 2 evaluates to `some_user` as the `ARG user` is defined on the
+subsequent line 3. The `USER` at line 4 evaluates to `what_user` as `user` is
+defined and `what_user` was passed on the command line. Prior to its definition by an
+`ARG` instruction, any use of a variable results in an empty string.
+
+> **Note:** It is not recommended to use build-time variables for
+>  passing secrets like github keys, user credentials etc.
+
+You can use an `ARG` or an `ENV` instruction to specify variables that are
+available to the `RUN` instruction. Environment variables defined using the
+`ENV` primitive of `Dockerfile` always override an `ARG` instruction of the same
+name. Consider this Dockerfile with an `ENV` and `ARG` instruction.
+
+```
+1 FROM ubuntu
+2 ARG CONT_IMG_VER
+3 ENV CONT_IMG_VER v1.0.0
+4 RUN echo $CONT_IMG_VER
+```
+Then, assume this image is built with this command:
+
+```
+$ docker build --build-arg CONT_IMG_VER=v2.0.1 Dockerfile
+```
+
+In this case, the `RUN` primitive uses `v1.0.0` instead of the `ARG` setting
+passed by the user:`v2.0.1` This behavior is similar to a shell
+script where a locally scoped variable overrides the variables passed as
+arguments or inherited from environment, from it's point of definition.
+
+Using the example above but a different `ENV` specification you can create more
+useful interactions between `ARG` and `ENV` instructions:
+
+```
+1 FROM ubuntu
+2 ARG CONT_IMG_VER
+3 ENV CONT_IMG_VER ${CONT_IMG_VER:-v1.0.0}
+4 RUN echo $CONT_IMG_VER
+```
+
+The command line passes the `--build-arg` and sets the `v2.0.1` value. And the `ARG
+CONT_IMG_VER` is defined on line 1 of the Dockerfile. On line 3, the `ENV`
+instruction of the same name resolves to to `v2.0.1` as the build-time variable
+was passed from command line and expanded here.
+
+The variable expansion technique in this example allows you to pass arguments
+from command-line and persist them in the final image by leveraging the `ENV`
+primitive. Variable expansion is only supported for the `Dockerfile` primitives
+described [here](#environment-replacement).
+
+Unlike an `ARG` instructions, `ENV` values are always persisted in the built image. If
+`docker build` were run without setting the `--build-arg` variable, then
+`CONT_IMG_VER` is still persisted in the image but its value would be `v1.0.0`.
+
+Finally, Docker has set of predefined `ARG` variables that you can use.
+
+* `HTTP_PROXY`
+* `HTTPS_PROXY`
+* `http_proxy`
+* `https_proxy`
+* `FTP_PROXY`
+* `NO_PROXY`
+
+To use these, simply pass them on the command line using the `--build-arg
+<varname>=<value>` flag. You can also define defaults for these in a Dockerfile.
+
```

Comment:
@thaJeztah 
> I cannot unset an ARG, so if a default value forhttp_proxy is defined in the Dockerfile, I cannot build the Dockerfile, unless I use a proxy.

Yes, that is correct.

The pre-defined args behave just like other args i.e. user may choose to define default value or description for them in the Dockerfile. But I agree in most common case users won't have/want defaults set for the current list of pre-defined args.

@moxiegirl 
> I'm cool with removing it as long as we state they otherwise behave as any other ARG meaning I can use it variable expansion etc

does the following wording bring out what you asked above?

```
Docker has a set of predefined `ARG` variables that you can use without a corresponding `ARG` instruction in the Dockerfile.
+
+* `HTTP_PROXY`
+* `HTTPS_PROXY`
+* `http_proxy`
+* `https_proxy`
+* `FTP_PROXY`
+* `NO_PROXY`
+
+To use these, simply pass them on the command line using the `--build-arg
+<varname>=<value>` flag.
```